nemo/collections/nlp/modules/common/hyena/hyena.py

```diff
@@ -0,0 +1,386 @@
+        def __init__(self, seqlen, dtype):
+            self.flashfftconv = FlashFFTConvImpl(seqlen, dtype)
+
+except ImportError:
+    HAVE_FLASHFFTCONV = False
+
+try:
+    from causal_conv1d import causal_conv1d_fn
+
+    HAVE_CAUSAL_CONV1D = True
+except ImportError:
+    HAVE_CAUSAL_CONV1D = False
+
+
+@dataclass
+class HyenaOperatorSubmodules:
+    in_proj: Union[ModuleSpec, type] = IdentityOp
+    short_filter: Union[ModuleSpec, type] = IdentityFuncOp
+    implicit_filter: Union[ModuleSpec, type] = IdentityOp
+    out_proj: Union[ModuleSpec, type] = IdentityOp
+
+
+def auto_assign_attrs(cls, **kwargs):
+    for k, v in kwargs.items():
+        setattr(cls, k, v)
+
+
+class CausalDepthWiseConv1d(nn.Module):
+    def __init__(self, channels, width, bias=True):
+        if not HAVE_CAUSAL_CONV1D:
+            raise ImportError("Missing causal-conv1d library, please run 'pip install causal-conv1d'")
+
+        super().__init__()
+        self.channels = channels
+        self.width = width
+        self._conv_1d = nn.Conv1d(
+            in_channels=channels,
+            out_channels=channels,
+            kernel_size=width,
+            padding=width - 1,
+            groups=channels,
+            bias=bias,
+        )
+
+    def forward(self, x):
+        return causal_conv1d_fn(x, self._conv_1d.weight.squeeze(1), self._conv_1d.bias)
+
+
+class HyenaConv(nn.Module):
+    def __init__(
+        self,
+        d_model: int,
+        l_max: int,
+        order: int,
+        filter_order: int,
+        bias: bool = True,
+        filter_cls: Union[ModuleSpec, type] = HyenaFilter,
+        filter_submodules: HyenaFilterSubmodules = None,
+        **filter_kwargs,
+    ):
+        super().__init__()
+        self.d_model = d_model
+        self.order = order
+        self.l_max = l_max
+        self.use_bias = bias
+        self.bias = nn.Parameter(torch.randn(self.d_model * (self.order - 1))) if self.use_bias else None
+
+        self.filter = build_module(
+            filter_cls,
+            self.d_model * (self.order - 1),
+            submodules=filter_submodules,
+            order=filter_order,
+            seq_len=l_max,
+            **filter_kwargs,
+        )
+
+
+class SingleHeadHyenaConv(HyenaConv):
+    def __init__(
+        self,
+        d_model: int,
+        l_max: int,
+        order: int,
+        filter_order: int,
+        bias: bool = True,
+        filter_cls: Union[ModuleSpec, type] = HyenaFilter,
+        filter_submodules: HyenaFilterSubmodules = None,
+        fftconv_type: str = None,
+        precision: str = 'bf16',
+        **filter_kwargs,
+    ):
+        super().__init__(
+            d_model,
+            l_max,
+            order,
+            filter_order,
+            bias=bias,
+            filter_cls=filter_cls,
+            filter_submodules=filter_submodules,
+            **filter_kwargs,
+        )
+
+        if fftconv_type is None:
+            if l_max <= 8192 and HAVE_FFTCONV:
```

Comment:
should we check if user uses l_max > 8k & fftconv_type = flash?

Reply:
Done